File: android/src/main/java/com/dcApi/DBModule.java
package com.dcApi;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import dcapi.Dcapi_;

/**
 * 数据库
 */
public class DBModule extends ReactContextBaseJavaModule {

    private static ReactApplicationContext reactContext;
    public Dcapi_ dcClass;

    public DBModule(ReactApplicationContext context, Dcapi_ dc) {
        super(context);
        reactContext = context;
        dcClass = dc;
    }

    @NonNull
    @Override
    public String getName() {
        return "DBModule";
    }



    // 创建数据库DB
    @ReactMethod
    public void db_NewThreadDB(
            String dbname,
            String rk,
            String sk,
            String jsonCollections,
            Callback successCallback,
            Callback errorCallback
    ){
        new Thread(new Runnable() {
            @Override
            public void run() {
                System.out.println("---------------------------------start newDB: " + dbname);
                String threadid = dcClass.db_NewThreadDB(dbname, rk, sk, jsonCollections);
                System.out.println("---------------------------------newDB threadid: " + threadid);
                if (threadid.equals("")) {
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------newDB: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                } else {
                    successCallback.invoke(threadid);
                }
            }
        }).start();
    }

    // syncDBFromDC 从DC网络中同步数据库信息到本地（一般发生在新设备首次登录时同步已经创建的数据库）,jsonCollections 是一个map结构的json字符串，格式[{"name":"name1","schema":"schema1"},indexs:[{"path":"path1","unique":true},{"path":"path2","unique":false}],{"name":"name2","schema":"schema2"},...]
    @ReactMethod
    public void db_SyncDBFromDC(
            String threadid,
            String dbname,
            String dbAddr,
            String rk,
            String sk,
            boolean block,
            String jsonCollections,
            Callback successCallback,
            Callback errorCallback
    ){
        new Thread(new Runnable() {
            @Override
            public void run() {
                System.out.println("---------------------------------start syncDBFromDC: " + threadid);
                Boolean bool = dcClass.db_SyncDBFromDC(threadid, dbname, dbAddr, rk, sk, block, jsonCollections);
                if (!bool) {
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------syncDBFromDC: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                } else {
                    successCallback.invoke();
                }
            }
        }).start();
    }


    // 同步数据库数据到本地
    @ReactMethod
    public void db_RefreshDBFromDC(
            String threadid,
            Callback successCallback,
            Callback errorCallback
    ){
        new Thread(new Runnable() {
            @Override
            public void run() {
                Boolean bool = dcClass.db_RefreshDBFromDC(threadid);
                System.out.println("---------------------------------Db_RefreshDBFromDC : " + bool);
                if (!bool) {
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------Db_RefreshDBFromDC: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                } else {
                    successCallback.invoke(true);
                }
            }
        }).start();
    }

    // 是否成功同步上传数据库DB
    @ReactMethod
    public void db_IfSyncDBToDCSuccess(
            String threadid,
            Callback successCallback,
            Callback errorCallback
    ){
        new Thread(new Runnable() {
            @Override
            public void run() {
                Boolean bool = dcClass.db_IfSyncDBToDCSuccess(threadid);
                System.out.println("---------------------------------ifSyncDBToDCSuccess : " + bool);
                if(!bool){
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------ifSyncDBToDCSuccess: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                }else {
                    successCallback.invoke(true);
                }
            }
        }).start();
    }

    // 同步上传数据库DB
    @ReactMethod
    public void db_SyncDBToDC(
            String threadid,
            Callback successCallback,
            Callback errorCallback
    ){
        new Thread(new Runnable() {
            @Override
            public void run() {
                Boolean bool = dcClass.db_SyncDBToDC(threadid);
                System.out.println("---------------------------------syncDBToDC : " + bool);
                if (!bool) {
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------syncDBToDC: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                } else {
                    successCallback.invoke(true);
                }
            }
        }).start();
    }

    // 通过threadid获取db
    @ReactMethod
    public void db_GetDBInfo(
            String threadid,
            Callback successCallback,
            Callback errorCallback
    ){
        new Thread(new Runnable() {
            @Override
            public void run() {
                System.out.println("---------------------------------start getDB: " + threadid);
                String jsonDBInfo = dcClass.db_GetDBInfo(threadid);
                if (jsonDBInfo.equals("")) {
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------getDB: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                } else {
                    successCallback.invoke(jsonDBInfo);
                }
            }
        }).start();
    }

    // 创建数据表记录
    @ReactMethod
    public void db_Create(
            String threadid,
            String collectionName,
            String jsonInstances,
            Callback successCallback,
            Callback errorCallback){
        new Thread(new Runnable() {
            @Override
            public void run() {
                String jsonInstanceids = dcClass.db_Create(threadid, collectionName, jsonInstances);
                System.out.println("---------------------------------create2: " + jsonInstanceids);
                if(jsonInstanceids.equals("")){
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------create: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                }else {
                    successCallback.invoke(jsonInstanceids);
                }
            }
        }).start();
    }

    // 删除数据表记录
    @ReactMethod
    public void db_Delete(
            String threadid,
            String collectionName,
            String instanceID,
            Callback successCallback,
            Callback errorCallback ){
        new Thread(new Runnable() {
            @Override
            public void run() {
                Boolean bool = dcClass.db_Delete(threadid, collectionName, instanceID);
                System.out.println("---------------------------------delete: " + bool);
                if(!bool){
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------delete: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                }else {
                    successCallback.invoke();
                }
            }
        }).start();
    }

    // 更新数据表记录
    @ReactMethod
    public void db_Save(
            String threadid,
            String collectionName,
            String instance,
            Callback successCallback,
            Callback errorCallback ){
        new Thread(new Runnable() {
            @Override
            public void run() {
                Boolean bool = dcClass.db_Save(threadid, collectionName, instance);
                System.out.println("---------------------------------saveInstance: " + threadid + "," + collectionName + "," + bool);
                if(!bool){
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------saveInstance: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                }else {
                    successCallback.invoke();
                }
            }
        }).start();
    }

    // 删除多条记录
    @ReactMethod
    public void db_DeleteMany(
            String threadid,
            String collectionName,
            String instanceIDs,
            Callback successCallback,
            Callback errorCallback ){
        new Thread(new Runnable() {
            @Override
            public void run() {
                Boolean bool = dcClass.db_DeleteMany(threadid, collectionName, instanceIDs);
                System.out.println("---------------------------------deleteManyInstance: " + bool);
                if(!bool){
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------deleteManyInstance: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                }else {
                    successCallback.invoke();
                }
            }
        }).start();
    }


    // 找到指定条件的数据表记录
    @ReactMethod
    public void db_Find(
            String threadid,
            String collectionName,
            String queryString,
            Callback successCallback,
            Callback errorCallback){
        new Thread(new Runnable() {
            @Override
            public void run() {
                String jsonInstances = dcClass.db_Find(threadid, collectionName, queryString);
                System.out.println("---------------------------------findInstance: " + jsonInstances);
                if(jsonInstances.equals("")){
                    String lastError = dcClass.dc_GetLastErr();
                    System.out.println("---------------------------------findInstance: err");
                    System.out.println(lastError);
                    errorCallback.invoke(lastError);
                }else {
                    successCallback.invoke(jsonInstances);
                }
            }
        }).start();
    }

//     // 查询某条数据记录
//     @ReactMethod
//     public void findByIDInstance(
//             String threadid,
//             String collectionName,
//             String instanceID,
//             Callback successCallback,
//             Callback errorCallback){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 try {
//                     String jsonInstance = dcClass.findByID(threadid, collectionName, instanceID);
//                     successCallback.invoke(jsonInstance);
//                     System.out.println("---------------------------------findByID: " + jsonInstance);
//                 } catch (Exception e) {
//                     errorCallback.invoke(e.getMessage());
//                     e.printStackTrace();
//                 }
//             };
//         };
//         Thread mt1 = new Thread(mt, "findByIDInstance");
//         mt1.start();
//     }

// //    // 监听变化
// //    @ReactMethod
// //    public void listenInstance(
// //            String threadid,
// //            String collectionName,
// //            String listenOptions,
// //            Callback successCallback,
// //            Callback callback){
// //        Runnable mt = new Runnable() {
// //            @Override
// //            public void run() {
// //                String handle = dcClass.listen(threadid, collectionName, listenOptions, new dc.Callback() {
// //                    @Override
// //                    public void inform(long l, String s, String s1) {
// //                        System.out.println("---------------------------------inform: ");
// //                        callback.invoke(l, s, s1);
// //                    }
// //                });
// //                System.out.println("---------------------------------listen1: " + handle);
// //                successCallback.invoke(handle);
// //                System.out.println("---------------------------------listen2: " + handle);
// //            };
// //        };
// //        Thread mt1 = new Thread(mt, "listenInstance");
// //        mt1.start();
// //    }


//     // 取消监听
//     @ReactMethod
//     public void cancelDbListen(
//             String handle
//     ){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 dcClass.cancelDbListen(handle);
//                 System.out.println("---------------------------------cancelDbListen: " + handle);
//             };
//         };
//         Thread mt1 = new Thread(mt, "cancelDbListen");
//         mt1.start();
//     }



// //    //将数据库内容添加到当前连接的dc peer（如果dc peer没有对当前用户授权，除非存储了该db内容的所有dc节点离线，否则将会添加失败）
// //    @ReactMethod
// //    public void addDbRecordToDcPeer(
// //            String threadid,
// //            String sk,
// //            Callback errorCallback
// //    ){
// //        try {
// //            dcClass.addDbRecordToDcPeer(threadid, sk);
// //            System.out.println("--------------------------------addDbRecordToDcPeer：");
// //        } catch (Exception e) {
// //            errorCallback.invoke(e.getMessage());
// //            e.printStackTrace();
// //        }
// //    }


//     // 返回该账号下的所有数据库
//     @ReactMethod
//     public void listDBs(
//             Callback successCallback,
//             Callback errorCallback
//     ){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 try {
//                     String jsonDBInfos = dcClass.listDBs();
//                     successCallback.invoke(jsonDBInfos);
//                     System.out.println("---------------------------------listDBs : " + jsonDBInfos);
//                 } catch (Exception e) {
//                     errorCallback.invoke(e.getMessage());
//                     e.printStackTrace();
//                 }
//             };
//         };
//         Thread mt1 = new Thread(mt, "listDBs");
//         mt1.start();
//     }

//     // 通过threadid删除db
//     @ReactMethod
//     public void deleteDB(
//             String threadid,
//             Callback successCallback,
//             Callback errorCallback
//     ){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 try {
//                     dcClass.deleteDB(threadid);
//                     System.out.println("---------------------------------deleteDB: ");
//                     successCallback.invoke();
//                 } catch (Exception e) {
//                     errorCallback.invoke(e.getMessage());
//                     e.printStackTrace();
//                 }
//             };
//         };
//         Thread mt1 = new Thread(mt, "deleteDB");
//         mt1.start();
//     }

//     // 创建数据表
//     // NewCollection creates a new collection.jsonconfig格式：{"name":"name1","schema":"schema1"},indexs:[{"path":"path1","unique":true},{"path":"path2","unique":false}]
//     @ReactMethod
//     public void newCollection(
//             String threadid,
//             String jsonconfig,
//             Callback successCallback,
//             Callback errorCallback ){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 try {
//                     dcClass.newCollection(threadid, jsonconfig);
//                     System.out.println("---------------------------------newCollection: ");
//                     successCallback.invoke();
//                 } catch (Exception e) {
//                     errorCallback.invoke(e.getMessage());
//                     e.printStackTrace();
//                 }
//             };
//         };
//         Thread mt1 = new Thread(mt, "newCollection");
//         mt1.start();
//     }

//     // 修改数据表
//     // UpdateCollection updates an existing collection.
//     @ReactMethod
//     public void updateCollection(
//             String threadid,
//             String jsonconfig,
//             Callback successCallback,
//             Callback errorCallback ){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 try {
//                     dcClass.updateCollection(threadid, jsonconfig);
//                     System.out.println("---------------------------------updateCollection: ");
//                     successCallback.invoke();
//                 } catch (Exception e) {
//                     errorCallback.invoke(e.getMessage());
//                     e.printStackTrace();
//                 }
//             };
//         };
//         Thread mt1 = new Thread(mt, "updateCollection");
//         mt1.start();
//     }

//     // 删除数据表
//     @ReactMethod
//     public void deleteCollection(
//             String threadid,
//             String name,
//             Callback successCallback,
//             Callback errorCallback ){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 try {
//                     dcClass.deleteCollection(threadid, name);
//                     System.out.println("---------------------------------deleteCollection: ");
//                     successCallback.invoke();
//                 } catch (Exception e) {
//                     errorCallback.invoke(e.getMessage());
//                     e.printStackTrace();
//                 }
//             };
//         };
//         Thread mt1 = new Thread(mt, "deleteCollection");
//         mt1.start();
//     }


//     // 获取数据库的所有集合
//     // ListCollections returns information about all existing collections.
//     @ReactMethod
//     public void listCollections(
//             String threadid,
//             Callback successCallback,
//             Callback errorCallback){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 try {
//                     String jsonConfig = dcClass.listCollections(threadid);
//                     successCallback.invoke(jsonConfig);
//                     System.out.println("---------------------------------listCollections: " + jsonConfig);
//                 } catch (Exception e) {
//                     errorCallback.invoke(e.getMessage());
//                     e.printStackTrace();
//                 }
//             };
//         };
//         Thread mt1 = new Thread(mt, "listCollections");
//         mt1.start();
//     }

//     // 获取数据表信息某个字段
//     @ReactMethod
//     public void getCollectionInfo(
//             String threadid,
//             String name,
//             Callback successCallback,
//             Callback errorCallback){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 try {
//                     String jsonConfig = dcClass.getCollectionInfo(threadid, name);
//                     successCallback.invoke(jsonConfig);
//                     System.out.println("---------------------------------getCollectionInfo: " + jsonConfig);
//                 } catch (Exception e) {
//                     System.out.println("---------------------------------getCollectionInfo error: " + e.getMessage());
//                     errorCallback.invoke(e.getMessage());
//                     e.printStackTrace();
//                 }
//             };
//         };
//         Thread mt1 = new Thread(mt, "getCollectionInfo");
//         mt1.start();
//     }

//     // 是否存在指定的数据表记录
//     @ReactMethod
//     public void hasInstance(
//             String threadid,
//             String collectionName,
//             String instanceIDs,
//             Callback successCallback ){
//         Runnable mt = new Runnable() {
//             @Override
//             public void run() {
//                 Boolean exist = dcClass.has(threadid, collectionName, instanceIDs);
//                 successCallback.invoke(exist);
//                 System.out.println("---------------------------------has: " + exist);
//             };
//         };
//         Thread mt1 = new Thread(mt, "hasInstance");
//         mt1.start();
//     }
}

File: ios/DcapiModules/DCApi/DBModule.m
//
//  DBModule.m
//  PrivatePhotos
//
//  Created by 王琴 on 2023/6/20.
//
#import "DBModule.h"
#import <React/RCTLog.h>//调用输出的方法
#import <React/RCTConvert.h>
#import <Dcapi/Dcapi.h>
#import "DcapiModules.h"


@interface DBModule ()
@end


@implementation DBModule
/**
/**
 *为了实现RCTBridgeModule协议，你的类需要包含RCT_EXPORT_MODULE()宏。这个宏也可以添加一个参数用来指定在 JavaScript 中访问这个模块的名字。如果你不指定，默认就会使用这个 Objective-C 类的名字。如果类名以 RCT 开头，则 JavaScript 端引入的模块名会自动移除这个前缀。
 */
RCT_EXPORT_MODULE()




#pragma mark - 注册方法addEvent
/**
 *你必须明确的声明要给 JavaScript 导出的方法，否则 React Native 不会导出任何方法。声明通过RCT_EXPORT_METHOD()宏来实现：
 *导出到 JavaScript 的方法名是 Objective-C 的方法名的第一个部分。React Native 还定义了一个RCT_REMAP_METHOD()宏，它可以指定 JavaScript 方法名。因为 JavaScript 端不能有同名不同参的方法存在，所以当原生端存在重载方法时，可以使用这个宏来避免在 JavaScript 端的名字冲突。
 *我们需要把事件的时间交给原生方法。我们不能在桥接通道里传递 Date 对象，所以需要把日期转化成字符串或数字来传递
 */
// 创建数据库DB
RCT_EXPORT_METHOD(db_NewThreadDB:(NSString*)dbname rk:(NSString*)rk sk:(NSString*)sk jsonCollections:(NSString*)jsonCollections  successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        RCTLogInfo(@"db_NewThreadDB");
        NSString *threadid = [dcapi db_NewThreadDB:dbname b32Rk:rk b32Sk:sk jsonCollections:jsonCollections];
        if(threadid.length > 0){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[threadid]);
             });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
             });
        }
    });
}

// syncDBFromDC 从DC网络中同步数据库信息到本地（一般发生在新设备首次登录时同步已经创建的数据库）,jsonCollections 是一个map结构的json字符串，格式[{"name":"name1","schema":"schema1"},indexs:[{"path":"path1","unique":true},{"path":"path2","unique":false}],{"name":"name2","schema":"schema2"},...]
RCT_EXPORT_METHOD(db_SyncDBFromDC:(NSString*)threadid dbname:(NSString*)dbname dbAddr:(NSString*)dbAddr rk:(NSString*)rk sk:(NSString*)sk block:(BOOL)block jsonCollections:(NSString*)jsonCollections  successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    RCTLogInfo(@"db_SyncDBFromDC");
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        BOOL success = [dcapi db_SyncDBFromDC:threadid dbname:dbname dbAddr:dbAddr b32Rk:rk b32Sk:sk block:block jsonCollections:jsonCollections];
        if(success){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[]);
            });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
            });
        }
    });
}

// 同步数据库数据到本地
RCT_EXPORT_METHOD(db_RefreshDBFromDC:(NSString*)threadid successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
  RCTLogInfo(@"db_RefreshDBFromDC");
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
      BOOL success = [dcapi db_RefreshDBFromDC:threadid];
      if(success){
          dispatch_async(dispatch_get_main_queue(), ^{
              successCallback(@[]);
           });
      }else {
          NSString *lastError = [dcapi dc_GetLastErr];
          dispatch_async(dispatch_get_main_queue(), ^{
              errorCallback(@[lastError]);
           });
      }
  });
}

// 是否成功同步上传数据库DB
RCT_EXPORT_METHOD(db_IfSyncDBToDCSuccess:(NSString*)threadid successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    RCTLogInfo(@"db_IfSyncDBToDCSuccess");
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        BOOL success = [dcapi db_IfSyncDBToDCSuccess:threadid];
        if(success){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[]);
            });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
            });
        }
    });
}

// 同步上传数据库DB
RCT_EXPORT_METHOD(db_SyncDBToDC:(NSString*)threadid successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    RCTLogInfo(@"db_SyncDBToDC");
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        BOOL success = [dcapi db_SyncDBToDC:threadid];
        if(success){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[]);
            });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
            });
        }
    });
}

// 通过threadid获取db
RCT_EXPORT_METHOD(db_GetDBInfo:(NSString*)threadid successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    RCTLogInfo(@"db_GetDBInfo");
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSString *jsonDBInfo = [dcapi db_GetDBInfo:threadid];
        if(jsonDBInfo.length > 0){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[jsonDBInfo]);
            });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
            });
        }
    });
}

// 创建数据表记录
RCT_EXPORT_METHOD(db_Create:(NSString*)threadid collectionName:(NSString*)collectionName jsonInstances:(NSString*)jsonInstances successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    RCTLogInfo(@"db_Create");
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSString *jsonInstanceids = [dcapi db_Create:threadid collectionName:collectionName jsonInstance:jsonInstances];
        if(jsonInstanceids.length > 0){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[jsonInstanceids]);
            });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
            });
        }
    });
}

// 删除数据表记录
RCT_EXPORT_METHOD(db_Delete:(NSString*)threadid collectionName:(NSString*)collectionName instanceID:(NSString*)instanceID successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    RCTLogInfo(@"db_Delete");
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        BOOL success = [dcapi db_Delete:threadid collectionName:collectionName instanceID:instanceID];
        if(success){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[]);
            });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
            });
        }
    });
}

// 更新数据表记录
RCT_EXPORT_METHOD(db_Save:(NSString*)threadid collectionName:(NSString*)collectionName instance:(NSString*)instance successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    RCTLogInfo(@"db_Save");
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        BOOL success = [dcapi db_Save:threadid collectionName:collectionName instance:instance];
        if(success){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[]);
            });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
            });
        }
    });
}

// 删除多条记录
RCT_EXPORT_METHOD(db_DeleteMany:(NSString*)threadid collectionName:(NSString*)collectionName instanceIDs:(NSString*)instanceIDs successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    RCTLogInfo(@"db_DeleteMany");
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        BOOL success = [dcapi db_DeleteMany:threadid collectionName:collectionName instanceIDs:instanceIDs];
        if(success){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[]);
            });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
            });
        }
    });
}

// 找到指定条件的数据表记录
RCT_EXPORT_METHOD(db_Find:(NSString*)threadid collectionName:(NSString*)collectionName queryString:(NSString*)queryString successCallback:(RCTResponseSenderBlock)successCallback errorCallback:(RCTResponseSenderBlock)errorCallback) {
    RCTLogInfo(@"db_Find");
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSString *jsonInstances = [dcapi db_Find:threadid collectionName:collectionName queryString:queryString];
        if(jsonInstances.length > 0){
            dispatch_async(dispatch_get_main_queue(), ^{
                successCallback(@[jsonInstances]);
            });
        }else {
            NSString *lastError = [dcapi dc_GetLastErr];
            dispatch_async(dispatch_get_main_queue(), ^{
                errorCallback(@[lastError]);
            });
        }
    });
}



#pragma mark - 向react-natvie 传递消息
- (NSArray<NSString *> *)supportedEvents
{
  return @[@"EventReminder"];
}
- (void)calendarEventReminderReceived:(NSNotification *)notification
{
  NSString *eventName = notification.userInfo[@"name"];
  [self sendEventWithName:@"EventReminder" body:@{@"name": eventName}];
}
- (dispatch_queue_t)methodQueue
{
  return dispatch_queue_create("com.facebook.React.AsyncLocalStorageQueue", DISPATCH_QUEUE_SERIAL);
}
-(id) init
{
    self = [super init];
    if(self)
    {
        if(dcapi == nil){
            [[DcapiModules alloc] init];
            NSLog(@"DBModule---------init---------========");
        }
    }
    return self;
}
@end
